src/graph.cc

```diff
@@ -252,14 +252,21 @@ bool DependencyScan::RecomputeOutputDirty(Edge* edge,
         EXPLAIN("command line changed for %s", output->path().c_str());
         return true;
       }
-      if (most_recent_input && entry->mtime < most_recent_input->mtime()) {
+      // Compute the effective mtime for the rule.
+      TimeStamp effective_mtime = entry->mtime;
+      if (generator) {
+        // For generator rules, always use the on-disk mtime since it is likely
+        // to have been regenerated outside of ninja.
+        effective_mtime = output->mtime();
```

Comment:
How can you be sure that `mtime()` doesn't return `-1`?

Reply:
Hmm. It is used above (line 220). I'll test out adding `&& output->status_known()` to the conditional guarding this.

Reply:
Still works with that conditional. I don't know exactly how the status wouldn't be known at this point, but it doesn't happen in typical CMake usage patterns at least.